### PR TITLE
Awaiting for topic subscription

### PR DESCRIPTION
**Pull request by wedens**
In some scenarios I'd like to know when specific subscriber has subscribed.

For example, I want to start a publisher and be sure that specific subscriber will receive all messages published by the publisher:

```scala
val topic: Topic[F, Msg] = ???
val messages: Stream[F, Msg] = ???
val publisher: Stream[F, Unit] = messages.to(topic.publish)
val messagesSink: Sink[F, Msg] = ???

val p = Stream.resource(topic.subscribeAwait(100)).flatMap { subscriber =>
  publisher concurrently subscriber.to(messagesSink)
}
```

Another possibility is to do something similar to `queue.dequeue.broadcastTo(sinks)` (with an assumption that you don't add subscribers between `Topic` creation and "starting" sinks):
```scala
val topic: Topic[F, Msg] = ???
val messages: Stream[F, Msg] = ???
val publisher: Stream[F, Unit] = messages.to(topic.publish)
val sinks: List[Sink[F, A]] = ???
val sinks0: Resource[F, List[Stream[F, Unit]]] =
  sinks.traverse(sink => topic.subscribeAwait(100).map(_.to(sink)))
val publisher: Stream[F, Unit] = ???

val p = Stream.resource(sinks0).flatMap { sinks1 =>
  publisher.concurrently(Stream.emits(sinks1).parJoinUnbounded.drain)
}
```
And it's still possible to add subscribers dynamically to the topic.

### CODE REVIEW

**Comment by wedens** [comment]
It breaks binary compatibility due to changed type of `imap` (`Resource.pure` requires `Applicative`). I'm not sure what to do with that.

**Comment by pchlupacek** [comment]
Ok so again a screwd with these comments. Had revire there already for few days before i realized i didnt hit submit :-(

File: core/shared/src/main/scala/fs2/concurrent/Topic.scala [hunk 1]
```diff
@@ -55,6 +55,17 @@ abstract class Topic[F[_], A] { self =>
...
+  /**
+    * Like [[subscribe]] but returns a "subscription" as a `Resource`.
+    *
+    * Useful when you need to know the point when specific "subscription" will start
+    * receiving values.
+    *
+    * @param maxQueued maximum number of elements to enqueue to the subscription
+    * queue before blocking publishers
+    */
+  def subscribeAwait(maxQueued: Int): Resource[F, Stream[F, A]]
```
[the remark this answers is not in the document]
Comment (wedens) [reply core/shared/src/main/scala/fs2/concurrent/Topic.scala]: "subscription" is a resource. Before this PR it was a singleton `Stream` which was `flatMap`ed in `subscribe`.

> I am not sure how this help you to 'to know the point when specific "subscription" will start'

When I do something with this Resource, I know that subscriber will receive messages from that point. With plain `subscribe` I don't know when it'll start receiving messages, so in `s.to(topic.publish) concurrently topic.subscribe(x).to(subscriber)` it's possible that `subscriber` will miss some messages.

**Comment by qingwei91** [comment]
Hi, is there anyway to move this forward?

**Comment by mpilquist** [comment]
Side note: we don't need the `Applicative` constraint on `imap` since we have a `Concurrent` constraint when building a `Topic`. For an example of this approach, see this commit: https://github.com/functional-streams-for-scala/fs2/commit/acfa259a37cf9f80c71e4013d19f112fe8db8ef1

However, both approaches still break binary compatibility as `subscribeAwait` is a new abstract method on the non-sealed `Topic` class.

We could target this for the upcoming 1.1 release which can have (small) binary incompatible changes.

**Comment by mpilquist** [comment]
@SystemFw @pchlupacek WDYT about this one?

**Comment by SystemFw** [comment]
There are a few decisions that we need to take about `Topic`, including what to do with https://github.com/functional-streams-for-scala/fs2/issues/1406.

> However, both approaches still break binary compatibility as subscribeAwait is a new abstract method on the non-sealed Topic class. We could target this for the upcoming 1.1 release which can have (small) binary incompatible changes.

Which scala versions are we causing bincompat breakage on?

**Comment by mpilquist** [comment]
See #1487 for binary compatibility issue -- basically, I think we release FS2 1.1 for 2.13 + cats-effect 2 and FS2 1.0 stays 2.11/2.12/cats-effect 1.

**Comment by SystemFw** [comment]
so fs2 1.1 won't be released for 2.12?

**Comment by mpilquist** [comment]
1.1 would still be cross-built for 2.11, 2.12, and 2.13

**Comment by SystemFw** [comment]
Right, so my question is: the breakage from this new op will be on 2.11 only or everywhere?

**Comment by mpilquist** [comment]
Oh I see where you're headed...

This new op breaks binary compat on 2.12 / 2.13 too, as it adds a new *abstract* method to an abstract class. Since `Topic` is an abstract class and not a trait, we *can* add concrete methods to it in a binary compatible way -- unlike if it was a trait, which runs in to the 2.11 trait encoding problem.

For this particular PR, I was thinking we'd schedule it such that it goes in to FS2 1.1.0, which will already be breaking binary compat due to cats-effect 2.0.0.

**Pull request closed without merging**